Python/ibmcloudsql/SQLQuery.py

```diff
@@ -31,99 +30,173 @@
+    from .catalog_table import HiveMetastore
+except Exception:
+    from catalog_table import HiveMetastore
+    
+import logging
+logger = logging.getLogger(__name__)
+from functools import wraps
+import json
+from json import JSONDecodeError
+import getpass
+import requests
+import dateutil.parser
+from pprint import pformat
+from collections import namedtuple
+import inspect
+import sqlparse
+
+from enum import Enum
+
+def check_saved_jobs_decorator(f):
+    """a decorator that load data from ProjectLib, check for completed SQL
+    Query job, before deciding to launch it"""
+    @wraps(f)
+    def wrapped(*args, **kwargs):
+        self = args[0]
+        dictionary = inspect.getcallargs(f, *args, **kwargs)
+        prefix = dictionary[
+            'prefix']  # Gets you the username, default or modifed
+        sql_stmt = dictionary["sql_stmt"]
+        # refine query
+        sql_stmt = format_sql(sql_stmt)
+
+        run_as_usual = True
+        if self.project_lib is not None:
+            # handle here
+            if self.project_lib.data is None:
+                self.read_project_lib_data(file_name=prefix)
+                keys_mapping = {}
+                # refine existing data
+                for key, _ in self.project_lib.data.items():
+                    new_key = format_sql(key)
+                    if key != new_key:
+                        keys_mapping[key] = new_key
+                if len(keys_mapping) > 0:
+                    rename_keys(self.project_lib.data, keys_mapping)
+
+            if sql_stmt in self.project_lib.data:
+                run_as_usual = False
+                job_id = self.project_lib.data[sql_stmt]["job_id"]
+                if self.project_lib.data[sql_stmt]["status"] == "completed":
+                    print("Job {} completed".format(job_id))
+                else:
+                    # query the status
+                    job_result = self.get_job(job_id)
+                    self.project_lib.data[sql_stmt]["job_info"] = job_result
+                    try:
+                        self.project_lib.data[sql_stmt]["status"] = job_result[
+                            "status"]
+                    except KeyError as e:
+                        import pprint
+                        pprint.pprint(job_result)
+                        raise e
+                    if job_result["status"] == "failed":
+                        run_as_usual = True
+
+        if run_as_usual:
+            job_id = f(*args, **kwargs)
+            if self.project_lib is not None:
+                self.project_lib.data[sql_stmt] = {
+                    "job_id": job_id,
+                    "status": "running"
+                }
+                self.write_project_lib_data()
+        return job_id
+
+    return wrapped
+
+    
+class SQLQuery(COSClient, SQLMagic, HiveMetastore):
+    """The class the provides necessary APIs to interact with
+
+    1. IBM SQL Serverless service
+    2. IBM COS service
+
+    Parameters
```

Comment:
max_tries is missing from the documentation

Reply:
updated